### PR TITLE
Add distinct counters to TSDB backends.

### PR DESCRIPTION
**Pull request by tkaemming**
This allows counting occurrences of unique items (e.g. user ID, URL, etc.) over
rollup intervals in a time range, as well as calculating the total number of
unique items over an aggregate time range that spans several rollup intervals.

The Redis implementation uses HyperLogLog to provide cardinality estimates (it
probably makes sense to document that there is a [approximate standard error of
0.81%](http://redis.io/commands/pfcount)) while the in-memory implementation
uses sets.

This adds support for the following time series queries:
- users that have been affected by events within a group (issue)
- users that have been affected by events within a project

### CODE REVIEW

File: src/sentry/tsdb/redis.py [hunk 1]
```diff
@@ -173,3 +174,74 @@ def get_range(self, model, keys, start, end, rollup=None):
         for key, points in results_by_key.iteritems():
             results_by_key[key] = sorted(points.items())
         return dict(results_by_key)
+
+    def record(self, model, key, values, timestamp=None):
+        self.record_multi(((model, key, values),), timestamp)
```
Comment (dcramer) [inline src/sentry/tsdb/redis.py]: maybe have the base class do this? idk

**Comment by dcramer** [comment]
Could you expand the documentation on the class docstring to include how the distinct counts architecture functions in Redis

File: src/sentry/tsdb/redis.py (hunk 1 above)
Comment (tkaemming) [inline src/sentry/tsdb/redis.py]: The base class does the opposite, this follows the same pattern established [here in `base.py`](https://github.com/getsentry/sentry/blob/2c6cd363bd5f85e685ddc03591529dc1002deb58/src/sentry/tsdb/base.py#L115-L130) and [in `redis.py`](
https://github.com/getsentry/sentry/blob/2c6cd363bd5f85e685ddc03591529dc1002deb58/src/sentry/tsdb/redis.py#L110-L134).

Comment (dcramer) [inline src/sentry/tsdb/redis.py]: Doh. Nevermind!

**Comment by tkaemming** [comment]
> Could you expand the documentation on the class docstring to include how the distinct counts architecture functions in Redis

Yeah, will do.

I also didn't implement this is in `sentry.tsdb.dummy.DummyTSDB`, since that doesn't implement the full API anyway. Is that even supported?

**Comment by dcramer** [comment]
By default it uses dummy TSDB so we should probably implement it there. It shouldnt ever be run, but its more of a No-Op backend.

File: src/sentry/tsdb/base.py [hunk 2]
```diff
@@ -44,6 +46,8 @@ class TSDBModel(Enum):
     # the number of events blocked due to being blacklisted
     organization_total_blacklisted = 202
 
+    users_affected_by_event = 300
```
Comment (tkaemming) [inline src/sentry/tsdb/base.py]: `event` should probably be `group` here.

**Comment by tkaemming** [comment]
The design in #2134 (specifically, "N issues affecting M users in the last X minutes" summary) line requires performing a cardinality estimate across *all events*. This could potentially cross nodes, and would require [porting the HyperLogLog count algorithm from the server implementation](https://github.com/antirez/redis/blob/3.0.4/src/hyperloglog.c) after retrieving the raw HLL representation from the server via `GET`.

**Comment by tkaemming** [comment]
I think this is basically ready to go (assuming it's greenlit by CI.)

**Pull request merged**